src/main/resources/static/MainPage.js

```diff
@@ -1,155 +1,223 @@
+                GetTable();
+                dialog_sub.dialog( "close" );
+            }
+            return valid;
+        }
+
+        /* Add perk for the subscription. */
+    function addPerk() {
+        var valid = true;
+
+        valid = valid && checkLength( name_perk, 3 );
+        valid = valid && checkLength( sub_perk, 3 );
+        valid = valid && checkLength( desc_perk, 3 );
+
+        if ( valid ) {
+            /* Construct JSON and send. This call is not async because the calls to GetTable finishes before this one.  */
+            var perkJson = {"name":name_perk.val(),"description":desc_perk.val(),"expiryDate":null,"subscription":null};
+            $.ajax({
+                type:"POST",
+                async:false,
+                contentType: "application/json; charset=utf-8",
+                url: cntxPath+"/AddPerk?userName="+userName+"&subName="+name_sub.val(),
+                dataType:"json",
+                data: JSON.stringify(perkJson)
+            });
+            /* Update table. */
+            GetTable();
+            dialog_perk.dialog( "close" );
+        }
+        return valid;
+    }
+
+    function addUpvoteListener(){
+        if ($('.upvotebutton').length == 0){
+            return;
+        }
+        $('.upvotebutton').click(function(event){
+
+            var subname = $(event.target).parent().parent().find("#subscription_name").text();
+            var perkname = $(event.target).parent().parent().find("#perk_name").text();
+
+            /* Construct JSON and send. This call is not async because the calls to GetTable finishes before this one.  */
+            var perkJson = {"name":perkname,"description":desc_perk.val(),"expiryDate":null,"subscription":null};
+            $.ajax({
+                type:"POST",
+                async:false,
+                contentType: "application/json; charset=utf-8",
+                url: cntxPath+"/upvote?userName="+userName+"&subName="+subname,
+                dataType:"json",
+                data: JSON.stringify(perkJson)
+            });
+            /* Update table. */
+            GetTable();
+            });
+    }
+
+    function addDownvoteListener(){
+        if ($('.downvotebutton').length == 0){
+            return;
+        }
+        $('.downvotebutton').click(function(event){
+
+            var subname = $(event.target).parent().parent().find("#subscription_name").text();
+            var perkname = $(event.target).parent().parent().find("#perk_name").text();
+
+            /* Construct JSON and send. This call is not async because the calls to GetTable finishes before this one.  */
+            var perkJson = {"name":perkname,"description":desc_perk.val(),"expiryDate":null,"subscription":null};
+            $.ajax({
+                type:"POST",
+                async:false,
+                contentType: "application/json; charset=utf-8",
+                url: cntxPath+"/downvote?userName="+userName+"&subName="+subname,

```

Comment:
You don't use either the userName or subName parameters in the controller.

Reply:
Done. Applied changes to PerkController.